rclcpp/include/rclcpp/parameter.hpp

```diff
@@ -0,0 +1,225 @@
+#include <rcl_interfaces/GetParameterTypes.h>
+#include <rcl_interfaces/Parameter.h>
+#include <rcl_interfaces/ParameterDescriptor.h>
+#include <rcl_interfaces/ParameterType.h>
+#include <rcl_interfaces/SetParameters.h>
+#include <rcl_interfaces/SetParametersAtomically.h>
+#include <rcl_interfaces/ListParameters.h>
+#include <rcl_interfaces/DescribeParameters.h>
+
+namespace rclcpp
+{
+
+namespace parameter
+{
+
+enum ParameterType {
+  PARAMETER_NOT_SET=rcl_interfaces::ParameterType::PARAMETER_NOT_SET,
+  PARAMETER_BOOL=rcl_interfaces::ParameterType::PARAMETER_BOOL,
+  PARAMETER_INTEGER=rcl_interfaces::ParameterType::PARAMETER_INTEGER,
+  PARAMETER_DOUBLE=rcl_interfaces::ParameterType::PARAMETER_DOUBLE,
+  PARAMETER_STRING=rcl_interfaces::ParameterType::PARAMETER_STRING,
+  PARAMETER_BYTES=rcl_interfaces::ParameterType::PARAMETER_BYTES,
+};
+
+// Structure to store an arbitrary parameter with templated get/set methods
+class ParameterVariant
+{
+public:
+  ParameterVariant()
+  : name_("")
+  {
+    value_.parameter_type = rcl_interfaces::ParameterType::PARAMETER_NOT_SET;
+  }
+  ParameterVariant(const std::string & name, const bool bool_value)
+  : name_(name)
+  {
+    value_.bool_value = bool_value;
+    value_.parameter_type = rcl_interfaces::ParameterType::PARAMETER_BOOL;
+  }
+  ParameterVariant(const std::string & name, const int64_t int_value)
+  : name_(name)
+  {
+    value_.integer_value = int_value;
+    value_.parameter_type = rcl_interfaces::ParameterType::PARAMETER_INTEGER;
+  }
+  ParameterVariant(const std::string & name, const double double_value)
+  : name_(name)
+  {
+    value_.double_value = double_value;
+    value_.parameter_type = rcl_interfaces::ParameterType::PARAMETER_DOUBLE;
+  }
+  ParameterVariant(const std::string & name, const std::string & string_value)
+  : name_(name)
+  {
+    value_.string_value = string_value;
+    value_.parameter_type = rcl_interfaces::ParameterType::PARAMETER_STRING;
+  }
+
+  /* Templated getter */
+  template<typename T>
+  T
+  get_value() const;
+
+  inline ParameterType get_typeID() const {return static_cast<ParameterType>(value_.parameter_type); }
+
+  inline std::string get_name() const {return name_; }
+
+  inline rcl_interfaces::ParameterValue get_parameter_value()
+  {
+    return value_;
+  }
+
+private:
+  std::string name_;
+  rcl_interfaces::ParameterValue value_;
+};
+
+template<>
+inline int64_t ParameterVariant::get_value() const
+{
+  if (value_.parameter_type != rcl_interfaces::ParameterType::PARAMETER_INTEGER) {
+    // TODO: use custom exception
+    throw std::runtime_error("Invalid type");
+  }
+  return value_.integer_value;
+}
+template<>
+inline double ParameterVariant::get_value() const
+{
+  if (value_.parameter_type != rcl_interfaces::ParameterType::PARAMETER_DOUBLE) {
+    // TODO: use custom exception
+    throw std::runtime_error("Invalid type");
+  }
+  return value_.double_value;
+}
+template<>
+inline std::string ParameterVariant::get_value() const
```

Comment:
Should we return `std::string` as `const &` even if it is not "symmetric" to the other gettters?